packages/aws-cdk/lib/api/garbage-collector.ts

```diff
@@ -0,0 +1,484 @@
+   * @default 'all'
+   */
+  readonly type: 's3' | 'ecr' | 'all';
+
+  /**
+   * The days an asset must be in isolation before being actually deleted.
+   *
+   * @default 0
+   */
+  readonly rollbackBufferDays: number;
+
+  /**
+   * The environment to deploy this stack in
+   *
+   * The environment on the stack artifact may be unresolved, this one
+   * must be resolved.
+   */
+  readonly resolvedEnvironment: cxapi.Environment;
+
+  /**
+   * SDK provider (seeded with default credentials)
+   *
+   * Will be used to make SDK calls to CloudFormation, S3, and ECR.
+   */
+  readonly sdkProvider: SdkProvider;
+
+  /**
+   * The name of the bootstrap stack to look for.
+   *
+   * @default DEFAULT_TOOLKIT_STACK_NAME
+   */
+  readonly bootstrapStackName?: string;
+
+  /**
+   * Max wait time for retries in milliseconds (for testing purposes).
+   *
+   * @default 60000
+   */
+  readonly maxWaitTime?: number;
+}
+
+/**
+ * A class to facilitate Garbage Collection of S3 and ECR assets
+ */
+export class GarbageCollector {
+  private garbageCollectS3Assets: boolean;
+  private garbageCollectEcrAssets: boolean;
+  private permissionToDelete: boolean;
+  private permissionToTag: boolean;
+  private bootstrapStackName: string;
+  private maxWaitTime: number;
+
+  public constructor(readonly props: GarbageCollectorProps) {
+    this.garbageCollectS3Assets = ['s3', 'all'].includes(props.type);
+    this.garbageCollectEcrAssets = ['ecr', 'all'].includes(props.type);
+
+    this.permissionToDelete = ['delete-tagged', 'full'].includes(props.action);
+    this.permissionToTag = ['tag', 'full'].includes(props.action);
+    this.maxWaitTime = props.maxWaitTime ?? 60000;
+
+    print(chalk.white(this.permissionToDelete, this.permissionToTag, props.action));
+
+    this.bootstrapStackName = props.bootstrapStackName ?? DEFAULT_TOOLKIT_STACK_NAME;
+
+    // TODO: ECR garbage collection
+    if (this.garbageCollectEcrAssets) {
+      throw new Error('ECR garbage collection is not yet supported');
+    }
+  }
+
+  /**
+   * Perform garbage collection on the resolved environment.
+   */
+  public async garbageCollect() {
+    print(chalk.black(this.garbageCollectS3Assets));
+    // SDKs
+    const sdk = (await this.props.sdkProvider.forEnvironment(this.props.resolvedEnvironment, Mode.ForWriting)).sdk;
+    const cfn = sdk.cloudFormation();
+    const s3 = sdk.s3();
+
+    const qualifier = await this.bootstrapQualifier(sdk, this.bootstrapStackName);
+
+    const activeAssets = new ActiveAssetCache();
+
+    let refreshStacksRunning = false;
+    const refreshStacks = async (isInitial?: boolean) => {
+      if (refreshStacksRunning) {
+        return;
+      }
+
+      refreshStacksRunning = true;
+
+      try {
+        const stacks = await this.fetchAllStackTemplates(cfn, qualifier);
+        for (const stack of stacks) {
+          activeAssets.rememberStack(stack);
+        }
+      } catch (err) {
+        throw new Error(`Error refreshing stacks: ${err}`);
```

Comment:
Also -- throwing an error would stop the background refresh (because we now do `setTimeout` so we don't reschedule ourselves), but it would not stop the collector, so we'll be collecting based on exceedingly stale information.